[9.1] ESQL: Prevent circular alias references in `DeduplicateAggs` (#139175)

kanoshiou opened the pull request:
# Backport

This will backport the following commits from `main` to `9.1`:
 - [ESQL: Prevent circular alias references in `DeduplicateAggs` (#139175)](https://github.com/elastic/elasticsearch/pull/139175)

<!--- Backport version: 10.2.0 -->

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

kanoshiou committed:
ESQL: Prevent circular alias references in `DeduplicateAggs` (#139175)

## Summary

Fixes circular reference detection in alias chains during
`DeduplicateAggs` optimizations.

## Problem

The `PushDownEnrich` optimization can create circular references when
pushing `Enrich` past `Project` operations, causing
`ql_illegal_argument_exception: Potential cycle detected` errors when
`DeduplicateAggs`.

Example query:

```esql
FROM languages_lookup
| RENAME language_name AS message
| ENRICH languages_policy ON language_code
| STATS a = count(to_lower(language_name)), b = count(language_name)
```

## Changes

Skip aliases that would create circular references by checking if
`alias.toAttribute()` equals the resolved `alias.child()`.

Closes #138346 Closes #139541

(cherry picked from commit 2289e1767a01a6df78babbde7f41eb43456dde9d)

# Conflicts:
#	x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRestTest.java
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich.csv-spec
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou commented:
Hi @bpintea,

Does this CI failure look familiar to you? I was wondering if you might have some insight.

bpintea commented:
Hey @kanoshiou,

> Does this CI failure look familiar to you? I was wondering if you might have some insight.

Unfortunately not. :)
I'd have to dig a bit to see what's going on.

kanoshiou committed:
Merge remote-tracking branch 'refs/remotes/origin/9.1' into backport/9.1/pr-139175

kanoshiou committed:
Merge remote-tracking branch 'origin/9.1' into backport/9.1/pr-139175

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich.csv-spec
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

alex-spies committed:
Merge remote-tracking branch 'upstream/9.1' into backport/9.1/pr-139175

alex-spies committed:
unmute GenerativeIT

alex-spies committed:
Fix tests

Lookup indices are not loaded in multi-cluster tests in 9.1. Instead of
languages_lookup we can also just use languages: same index, just
without the lookup mode set.

alex-spies committed:
Revert "Fix tests"

This reverts commit b0696b52ba9faa06f7afb4d0763c133fcd673f0c.

Turns out, we cannot use the languages index because this one is also
used for enrich, and the multi cluster tests index this two times, on
the remote AND on the local cluster.

alex-spies committed:
Use index that isn't duplicated in multi-cluster

alex-spies committed:
Fix more tests...

alex-spies committed:
Merge branch '9.1' into backport/9.1/pr-139175
